File: backend/types/economy.go
package types

var EconomyTemplate = *loadConfig[Economy]()

type Economy struct {
	Businesses []Business `yaml:"Businesses"`
	Resources  []Resource `yaml:"Resources"`
}

type Business struct {
	BusinessId   `yaml:"Id"`
	Icon         Icon                       `yaml:"Icon"`
	Description  string                     `yaml:"Description"`
	CanBeHobby   bool                       `yaml:"CanBeHobby"`
	Requirements []ProductionFlowBatchInput `yaml:"Requirements"`
	Productions  []Production               `yaml:"Productions"`
}

type Resource struct {
	ResourceID `yaml:"Id"`
	Icon       Icon `yaml:"Icon"`
	Category   Tag  `yaml:"Category"`
}

type Production struct {
	ProductionId     `yaml:"Id"`
	Icon             Icon              `yaml:"Icon"`
	UserSkillEffects []UserSkillEffect `yaml:"UserSkillEffects"`
	Flows            []ProductionFlow  `yaml:"Flows"`
}

type ProductionFlow struct {
	ProductionFlowId `yaml:"Id"`
	MaxPeople        int                           `yaml:"MaxPeople"`
	Input            []ProductionFlowBatchInput    `yaml:"Input"`
	Catalysts        []ProductionFlowBatchCatalyst `yaml:"Catalysts"`
	Output           []ProductionFlowBatch         `yaml:"Output"`
}

type ProductionFlowBatch struct {
	ResourceID    `yaml:"Id"`
	AmountPerTurn float32 `yaml:"AmountPerTurn"`
}

type ProductionFlowBatchInput struct {
	ProductionFlowBatch `yaml:",inline"`
	IsConsumed          bool `yaml:"IsConsumed"`
}

type ProductionFlowBatchCatalyst struct {
	ProductionFlowBatchInput `yaml:",inline"`
	OutputMultiplier         map[string]float32 `yaml:"OutputMultiplier"`
	OutputAdd                map[string]float32 `yaml:"OutputAdd"`
	InputMultiplier          map[string]float32 `yaml:"InputMultiplier"`
	InputSubtract            map[string]float32 `yaml:"InputSubtract"`
}

File: frontend/src/types/economy.ts
import { UserSkillEffect } from './effect';

export interface Economy {
  Businesses: Business[];
  Resources: Resource[];
}

export interface Business {
  BusinessId: string;
  Icon: string;
  Description: string;
  CanBeHobby: boolean;
  Requirements: ProductionFlowBatchInput[];
  Productions: Production[];
}

export interface ResourceCategory {
  CategoryId: string;
  Icon: string;
}

export interface Resource {
  ResourceId: string;
  Icon: string;
  CategoryId: string;
}

export interface Production {
  ProductionId: string;
  Icon: string;
  UserSkillEffects: UserSkillEffect[];
  Flows: ProductionFlow[];
}

export interface ProductionFlow {
  FlowId: string;
  MaxPeople: number;
  Input: ProductionFlowBatchInput[];
  Catalysts: ProductionFlowBatchCatalyst[];
  Output: ProductionFlowBatch[];
}

export interface ProductionFlowBatch {
  BatchId: string;
  AmountPerTurn: number;
}

export interface ProductionFlowBatchInput extends ProductionFlowBatch {
  IsConsumed: boolean;
}

export interface ProductionFlowBatchCatalyst extends ProductionFlowBatchInput {
  OutputMultiplier: Record<string, number>;
  OutputAdd: Record<string, number>;
  InputMultiplier: Record<string, number>;
  InputSubtract: Record<string, number>;
}
